### DISCUSSION TITLE
Repository priority modifier "include" to complement "exclude" and "only"?

### DISCUSSION THREAD

**Discussion by dzuelke** [composer/composer#10505] [General]
Aside from `canonical`, there are only two controls for [repository priorities](https://getcomposer.org/doc/articles/repository-priorities.md) right now to control individual packages:
- `exclude`, which filters out given packages from a repo on load
- `only`, which filters out all except the given packages from a repo on load

What's not possible, right now, is to specify a repository that's allowed to load certain packages, even if a previous canonical repository already contains these packages.

The workaround is to either set the previous repository to `"canonical": false`, which is probably not the best practice, or to set `"exclude":["somepackage"]` on the previous repository, to allow the next repository to load the package.

The trouble with the first workaround is that it's not really a recommended practice to set a repo to non-canonical; the second workaround has the problem that the declaration on what package should be loadable in repo B must be set on repo A.

Is the lack of `include` as a priority modifier intentional, or should I look into making a PR?

**Comment by Seldaek** [comment]
Hmm, it's just going to be very messy to implement I think given how it's been done. This is all self contained in https://github.com/composer/composer/blob/main/src/Composer/Repository/FilterRepository.php which wraps other repos. Once a repo returns that it has loaded a package, the following ones are not checked at all.

The way I understand what you want, `include` would have to transcend this and coordinate what gets fetched from all repos together in a different way.

And yes your two workarounds are all options I see right now.
